flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderBase.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.leaderelection;
+
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+/**
+ * Base class which provides some convenience functions for testing purposes of {@link LeaderContender} and
+ * {@link LeaderElectionEventHandler}.
+ */
+public class TestingLeaderBase {
+
+	protected boolean leader = false;
+	protected Throwable error = null;
+
+	protected final Object lock = new Object();
+	private final Object errorLock = new Object();
+
+	/**
+	 * Waits until the contender becomes the leader or until the timeout has been exceeded.
+	 *
+	 * @param timeout
+	 * @throws TimeoutException
+	 */
+	public void waitForLeader(long timeout) throws TimeoutException {
+		waitFor(this::isLeader, timeout, "Contender was not elected as the leader within " + timeout + "ms");
+	}
+
+	/**
+	 * Waits until the contender revokes the leader or until the timeout has been exceeded.
+	 *
+	 * @param timeout
+	 * @throws TimeoutException
+	 */
+	public void waitForRevokeLeader(long timeout) throws TimeoutException {
+		waitFor(() -> !isLeader(), timeout, "Contender was not revoked within " + timeout + "ms");
+	}
+
+	protected void waitFor(Supplier<Boolean> supplier, long timeout, String msg) throws TimeoutException {
+		long start = System.currentTimeMillis();
+		long curTimeout;
+
+		while (!supplier.get() && (curTimeout = timeout - System.currentTimeMillis() + start) > 0) {
+			synchronized (lock) {
+				try {
+					lock.wait(curTimeout);
+				} catch (InterruptedException e) {
+					// we got interrupted so check again for the condition
+				}
+			}
+		}
+
+		if (!supplier.get()) {
+			throw new TimeoutException(msg);
+		}
```

Comment:
I am wondering whether we can't solve this problem a bit more elegantly using a blocking array queue.

Reply:
Hmm. Actually, this is just the old codes I moved from `TestingContender`. It is indeed could be optimized. I will do it.

Reply:
This is not strictly needed right now. We could tackle this as part of a follow up issue.